stdlib/camlinternalQuote.ml

```diff
@@ -1196,502 +1292,781 @@ module Ast = struct
-    let heap, argo = fold_map_option merge loc Level_map.empty argo in
-      mk loc heap (Pexp_construct (lid, argo))
-
-  let variant loc label argo =
-    let heap, argo = fold_map_option merge loc Level_map.empty argo in
-      mk loc heap (Pexp_variant (label, argo))
-
-  let record loc defs orig =
-    let heap, defs = fold_left_map_alist merge loc Level_map.empty defs in
-    let heap, orig = fold_map_option merge loc heap orig in
-      mk loc heap (Pexp_record (defs, orig))
-
-  let field loc rcd lid =
-    let heap, rcd = merge loc Level_map.empty rcd in
-      mk loc heap (Pexp_field (rcd, lid))
-
-  let setfield loc rcd lid def =
-    let heap, rcd = merge loc Level_map.empty rcd in
-    let heap, def = merge loc heap def in
-      mk loc heap (Pexp_setfield (rcd, lid, def))
-
-  let array loc args =
-    let heap, args = List.fold_left_map merge loc Level_map.empty args in
-      mk loc heap (Pexp_array args)
-
-  let ifthenelse loc cond tr fs =
-    let heap, cond = merge loc Level_map.empty cond in
-    let heap, tr = merge loc heap tr in
-    let heap, fs = fold_map_option merge loc heap fs in
-      mk loc heap (Pexp_ifthenelse (cond, tr, fs))
-
-  let sequence loc exp1 exp2 =
-    let heap, exp1 = merge loc Level_map.empty exp1 in
-    let heap, exp2 = merge loc heap exp2 in
-      mk loc heap (Pexp_sequence(exp1, exp2))
-
-  let while_ loc cond body =
-    let heap, cond = merge loc Level_map.empty cond in
-    let heap, body = merge loc heap body in
-      mk loc heap (Pexp_while(cond, body))
-
-  let for_nonbinding loc pat low high dir body =
-    let dir = if dir then CamlinternalAST.Upto else CamlinternalAST.Downto in
-    let pat = PatRepr.nonbinding loc pat in
-    let heap, low = merge loc Level_map.empty low in
-    let heap, high = merge loc heap high in
-    let heap, body = merge loc heap body in
-      mk loc heap (Pexp_for (pat, low, high, dir, body))
-
-  let for_simple loc name low high dir f =
-    let dir = if dir then CamlinternalAST.Upto else CamlinternalAST.Downto in
-    let heap, pat, body = Binding.simple loc name f in
-    let heap, low = merge loc heap low in
-    let heap, high = merge loc heap high in
-      mk loc heap (Pexp_for (pat, low, high, dir, body))
-
-  let assert_ loc exp =
-    let heap, exp = merge loc Level_map.empty exp in
-      mk loc heap (Pexp_assert exp)
-
-  let lazy_ loc exp =
-    let heap, exp = merge loc Level_map.empty exp in
-      mk loc heap (Pexp_lazy exp)
-
-  let open_ loc ovr lid exp =
-    let ovr =
-      if ovr then CamlinternalAST.Override else CamlinternalAST.Fresh
+    With_free_vars.value_bindings loc names (fun vars ->
+        let pat, body = f vars in
+        let+ pat, body = With_free_vars.both pat body and+ def = def in
+        let pat =
+          With_bound_vars.value
+            ~extra:(BindingError.unexpected_binding_error_at_loc loc)
+            ~missing:BindingError.missing_binding_error pat
+            (List.map Var.Value.generic vars)
```

Comment:
It looks like you copied this from some of my code, but reading it now it feels a little silly to map the identity function over the list. Maybe add a `Var.Value.generic_list` function to do that without runtime cost.